L70-node-proto-loader-type-generator.md

```diff
@@ -0,0 +1,237 @@
+@grpc/proto-loader TypeScript Type Generator CLI Tool
+----
+* Author(s): murgatroid99
+* Approver: wenbozhu
+* Status: Draft
+* Implemented in: Node.js
+* Last updated: 2020-06-01
+* Discussion at: https://groups.google.com/g/grpc-io/c/FwLprMC-5Qc
+
+## Abstract
+
+Add a tool to the `@grpc/proto-loader` to generate types that describe the objects that will be generated by loading the output of `@grpc/proto-loader` into `grpc` or `@grpc/grpc-js`.
+
+## Background
+
+`@grpc/proto-loader` outputs objects with types determined by the `.proto` files loaded at runtime, so it is currently very difficult to get compile-time type information about those objects. As TypeScript becomes increasingly popular, that compile-time type information becomes increasingly desirable.
+
+
+### Related Proposals: 
+* L23 Standalone Node.js gRPC+Protobuf.js Library API
+* L43 Node Message Type Information
+
+## Proposal
+
+In the `@grpc/proto-loader` library, provide a command line tool that will generate the types of the objects that will be created by passing the output of `load` or `loadSync` to `grpc.loadPackageDefinition` when loading a specific set of `.proto` files with a specific set of options. This will allow users to write code like this to use the type information in the rest of the code:
+
+```ts
+import * as grpc from 'grpc';
+// OR 
+import * as grpc from '@grpc/grpc-js';
+
+import { ProtoGrpcType, ServiceHandlers, messages } from 'path/to/proto-file-name_proto'; // File generated by the tool is path/to/proto-file-name_proto.d.ts
+import * as protoLoader from '@grpc/proto-loader';
+
+const packageDefinition = protoLoader.loadSync('other/path/to/proto-file-name.proto', options);

```

Comment:
It's quite common for folks to want to load more than one proto file. It would be good to switch this over to an array instead of a simple string to show how multiple types may be loaded.

Reply:
This API does accept an array of strings as an alternative to a single string. When loading multiple files you can get the type of the output from `grpc.loadPackageDefinition` by loading all of the corresponding `ProtoGrpcType` interfaces from the corresponding generating files and combining them with `&`.

Also, maybe this isn't clear, but the proposed CLI tool can also accept multiple filenames to generate types for.